[DOC] Adds developer guideline start

jbteves opened the pull request:
<!---
This is a suggested pull request template for tedana.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

Please also label your pull request with the relevant tags.
See here for more information and a list of available options:
https://github.com/ME-ICA/tedana/blob/master/CONTRIBUTING.md#pull-requests
-->

<!-- Please indicate after the # which issue you're closing with this PR.
This is helpful for the maintainers AND will magically close the issue when this
pull request is merged!
https://help.github.com/articles/closing-issues-using-keywords -->
Closes #268, #479 

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- Adds Developer Guidelines to RTD
- Moves monthly call to RTD
- Adds worked example to RTD
- Adds more information on git branching to CONTRIBUTING
- Updates some information in CONTRIBUTING
- Increases required reviewer count to 2

emdupre commented on CONTRIBUTING.md:
````diff
@@ -159,22 +144,30 @@ Before creating your pull request, please make sure to review the `tedana` [styl
-(This does require that you have a local install of [Docker](https://www.docker.com/products/docker-desktop).)
-You can run all the checks with:
+```
+cd TEDANADIR
````
Can you explain that `TEDANADIR` is a file path, here ? And, can we mark it as a variable with a `$` ?

emdupre commented on CONTRIBUTING.md:
````diff
@@ -159,22 +144,30 @@ Before creating your pull request, please make sure to review the `tedana` [styl
+and unit/integration tests by running the following:
 ```
-docker run --tty --rm -v ${PWD}:/tedana tedana/tedana-dev:latest run_all_tests
+cd TEDANADIR/tedana/tests
````
I don't think you need to navigate into the directory to run pytest. It should find them just from the $TEDANADIR  with `pytest tedana`.

emdupre commented on CONTRIBUTING.md:
````diff
@@ -159,22 +144,30 @@ Before creating your pull request, please make sure to review the `tedana` [styl
+flake8 tedana
+```
 
+and unit/integration tests by running the following:
````
Since you give a number of options below, can we either add more explanatory text here or break up that block to reference the different calls separately ?

emdupre commented on CONTRIBUTING.md:
```diff
@@ -159,22 +144,30 @@ Before creating your pull request, please make sure to review the `tedana` [styl
-
-This will print out a number of different status update messages as the tests run, but if you see `"FINISHED RUNNING ALL TESTS! GREAT SUCCESS"` then it means everything finished succesfully.
-If not, there should be some helpful outputs that specify which tests failed.
+The test run will indicate the number of passes and failures.
```
Should we explain these are run as part of the automated testing, too ?

emdupre commented on CONTRIBUTING.md:
```diff
@@ -202,20 +195,28 @@ When opening a pull request, please use at least one of the following prefixes:
-We request that you do not use the Draft PR feature at this time, 
-as it interferes with our Continuous Integration tool, Travis.
+Note that if your pull request has no conversation or commits for 90 days,
+stale-bot will mark it stale and we may ask your permission to take over it.
```
```suggestion
stale-bot will mark it stale.
We may also ask that another development team member resume work on it, if you are unable to continue to do so.
```

emdupre commented on CONTRIBUTING.md:
```diff
@@ -202,20 +195,28 @@ When opening a pull request, please use at least one of the following prefixes:
-You can also combine the tags above, for example if you are updating both a test and
-the documentation: **[TST, DOC]**.
-If you're still working on the pull request that prefix would be **[WIP, TST, DOC]**.
+Pull Request Checklist (Before Requesting Review):
```
Can this be a heading ?

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,155 @@
+Developer Guidelines
+====================
+
+Git Workflow
```
I really think this should go back to the CONTRIBUTING...

jbteves replied:
The `git` workflow or the whole page?

emdupre commented on docs/developing.rst:
Same concern about `TEDANADIR`, here.

jbteves commented on CONTRIBUTING.md:
It was originally intended for use on the RTD, which is more removed from new users, but good point now that it's in CONTRIBUTING

jbteves commented on docs/developing.rst:
```diff
@@ -0,0 +1,87 @@
+    curl https://circleci.com/api/v1.1/project/:vcs-type/:username/:project/$build_number/artifacts?circle-token=$CIRCLE_TOKEN \
+       | grep -o 'https://[^"]*' \
+       | sed -e "s/$/?circle-token=$CIRCLE_TOKEN/" \
+       | wget -v -i -
```
Addressed in d722317

jbteves commented on docs/developing.rst:
```diff
@@ -0,0 +1,87 @@
+
+For integration tests,
+make a ``tar.gz`` file which will unzip to be a single directory
+containing all of the files you'd like to run a workflow on.
```
I meant to type "archive," whoops. Addressed in a63ccfe

jbteves commented on CONTRIBUTING.md:
```diff
@@ -33,15 +32,6 @@ We also maintain a [gitter chat room][link_gitter] for more informal conversatio
-
-Everyone is welcome.
-We look forward to meeting you there :hibiscus:
-
```
Yes, I must have somehow gotten distracted and not pasted it in. Addressed in aa41ffe, with a slight modification since the agenda is often last-minute.

jbteves commented:
@rmarkello  
Example: Yes, I think I had originally intended to add it and then got so excited that I finished some other bits that I forgot about it. My current plan is to close #252 and use it as an example, unless somebody has an alternative suggestion. 
MD file: We can change that if you like. I think @KirstieJane would also like that. I don't know what conventions others have, but mine is breaks at commas, periods, and 75 characters (editor width). 
RST file: Breaks follow the same convention as MD file for me. I can adjust, and did so in the commits I just pushed. 
The comma-period-75 rule is a habit from running lots of LaTeX with vim because MacTex was (is?) abysmally slow, but the underlying CLI wasn't (isn't?). Sorry about that, I'm sure it looks insane to everyone else using fancy new editors. 
There's nothing in any style guide here about line widths as far as I can tell, and we tell people to use PEP8 even though flake8 is set to 99; should we add that, too?
I don't have an issue with long line lengths in marked-up text such as MD and RST, however...
I think that 99 characters for code is incredibly long. 75-80 is much more standard because 80 is the standard terminal width and some people like to use smaller widths in case editors with line numbers are used (like me). It's also not PEP8 compliant, since the PEP8 standard is 79.
Tagging @emdupre and @KirstieJane to weigh on line length.

jbteves commented:
We could also use `doc8` to enforce style, however the default is 80.

emdupre commented:
> I think that 99 characters for code is incredibly long. 75-80 is much more standard because 80 is the standard terminal width and some people like to use smaller widths in case editors with line numbers are used (like me). It's also not PEP8 compliant, since the PEP8 standard is 79.

I know :smile_cat:  But the 99 is intentional, here -- we added it to the setup.cfg to override that PEP8 error. Particularly in some of the documentation (and looking through Dan's new decision tree), we just hit 79 characters too often. The extra 20 characters ends up being really valuable to keep the code readable. Though, admittedly, we're geared towards the browser or VSCode. 

I'd like to keep MD and RST on the same line length, though I do generally recommend putting each sentence on a new line !

jbteves commented:
I still feel like it's too long but it seems that the consensus is that 99 is requireds, so I'll work with it.
It will take an extremely long time to fix that in the RST; there's about 800 doc8 errors from whitespace and line lengths. For some reason I can't seem to get the linting configured in VSCode, so the next best thing is enforcing it in vim and re-running doc8, but that tends to break the formatting, so I have to go back over it manually to correct. 
I also went ahead and added doc8 to the requirements and set it up to parse anything that isn't in `/_build`.

emdupre commented:
If it takes too long I wouldn't do that here, then -- it'll bring up a lot of changes that will make this harder to review. And I want to make sure this gets a thorough review, since it's really important !!

jbteves commented:
Per in-person conversation with @emdupre, I split off the changes to `developing.rst` and basically I made a stash of changes so far for other formatting that I can just apply later.

jbteves commented:
I also decided #252 might be too difficult to do as an example, so I'm switching to a totally fake one unless someone objects.

jbteves commented:
@emdupre @rmarkello check out the worked example here: https://external-builds.readthedocs.io/html/tedana/446/developing.html#worked-example

jbteves commented:
I think that this is ready for re-view @emdupre @rmarkello, hopefully this one's the charm or close to it.

rmarkello commented on CONTRIBUTING.md:
```diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
-This will print out a number of different status update messages as the tests run, but if you see `"FINISHED RUNNING ALL TESTS! GREAT SUCCESS"` then it means everything finished succesfully.
-If not, there should be some helpful outputs that specify which tests failed.
+The test run will indicate the number of passes and failures.
+Most often, the failures give enough information to determine the cause; if not, if not you can
```
```suggestion
Most often, the failures give enough information to determine the cause; if not, you can
```

rmarkello commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+    tar czf my_data.tar.gz my_data/*.nii.gz
+
+Run the workflow with a known-working version, and put the outputs into a text file inside
+``$TEDANADIR/tedana/tests/data/``,with ``TEDANADIR`` the local ``tedana repository``.
```
```suggestion
``$TEDANADIR/tedana/tests/data/``, where ``TEDANADIR`` is your local ``tedana repository``.
```

rmarkello commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+
+Run the workflow with a known-working version, and put the outputs into a text file inside
+``$TEDANADIR/tedana/tests/data/``,with ``TEDANADIR`` the local ``tedana repository``.
+You can follow the model our `five echo set`_, which has the following steps:
```
```suggestion
To write the test function you can follow the model of our `five echo set`_, which takes the following steps:
```

rmarkello commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+   actual output.
+
+If you need to upload new data, you will need to contact the maintainers and ask them to either add
+it or give you permission to add it.
```
```suggestion
it to `OSF`_ or give you permission to add it.
```

And then can you add the `OSF`_ link to the links at the bottom?

rmarkello commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+You can see the schedule via the tedana `google calendar`_.
+
+Everyone is welcome.
+We look forward to meeting you there :hibiscus:
```
```suggestion
We look forward to meeting you there!
```

The emoji aren't supported on .rst, unfortunately 😢

jbteves replied:
Even worse, there is no sad hibiscus emoji for me to respond with.

emdupre commented:
@KirstieJane and @javiergcas -- I requested your review here since we've talked separately on improving the developer docs ! Would be great to get your feedback :sparkles:

handwerkerd commented:
In the current CONTRIBUTING.md and this PR, there is the  following sentence. "Here's one guide we've found particularly helpful for starting with rst." That links to http://docs.sphinxdocs.com/en/latest/step-1.html which now looks like a domain squatter.  
Could someone add a link to a 'starting with rst' guide that they link to this PR?

jbteves commented:
This isn't a getting started guide, but a reference. WDYT @handwerkerd ?
http://docutils.sourceforge.net/docs/user/rst/quickref.html

handwerkerd commented:
@jbteves That link is already included in the previous sentence (or at least http://docutils.sourceforge.net/rst.html#user-documentation is, which links to that fairly clearly). If what ever guide something thought was good can be found elsewhere, there should be a fixed link. If nothing else is obviously worth linking here, then just remove the now dead link

rmarkello commented:
This guide from the Sphinx Docs is pretty good: http://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html! Perhaps that can be linked instead of the dead link :+1:

handwerkerd commented:
http://www.sphinx-doc.org/en/master/index.html is already in the same paragraph as well. Perhaps replace the specific sphinx & docutils links with these more specific pages on each site & remove the dead link.

emdupre commented on CONTRIBUTING.md:
```diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
 
-Before creating your pull request, please make sure to review the `tedana` [style conventions](#style-guide).
 
 ### 5. Test your changes
```
One general thing -- I think the way this section is currently written risks discouraging people from opening pull requests "early and often" which is something we explicitly want.

Especially since the testing is so minimal, I'd almost rather we didn't include this section and instead wrote something about how we expect all new code to include tests, and guides for writing good tests. WDYT ?

jbteves replied:
That sounds like a good alternative. I can see why the phrasing I had is discouraging, I'll amend it ahead of Thanksgiving.

jbteves replied:
@emdupre do you consider this resolved now since you approved earlier?

emdupre commented on CONTRIBUTING.md:
```diff
@@ -202,20 +237,45 @@ When opening a pull request, please use at least one of the following prefixes:
+stale-bot will mark it stale.
+We may also ask that another development team member resume work on it, if you are unable to continue to do so.
+
+### Pull Request Checklist (Before Requesting Review):
```
Again, we want to make sure we're not discouraging people from _opening_ pull requests, just setting expectations around what will get a full review.

emdupre commented on CONTRIBUTING.md:
```diff
@@ -202,20 +237,45 @@ When opening a pull request, please use at least one of the following prefixes:
+
+### After Changes Are Merged
+After successful merging of the pull request, remember to [keep your fork up to date][link_updateupstreamwiki] with the master `tedana` repository and to delete the branch on your fork that was used for the merged pull request.
+You may also want to update your other branches to include these changes,
```
I think this is redundant with the idea of merging in master, which I'd again encourage folks not to use unless there's a merge conflict.

rmarkello replied:
Agreed. Since we're suggesting as a first step in the contribution process that users fetch & merge upstream/master (before creating a new branch), I think we can safely remove this section!

jbteves replied:
Okay, I've just deleted those sections.

emdupre commented on CONTRIBUTING.md:
````diff
@@ -202,20 +237,45 @@ When opening a pull request, please use at least one of the following prefixes:
+git merge master
+```
+
+In the event of confusion, please ping in the Gitter for help.
````
Let's not recommend Gitter, here.

jbteves replied:
🤣 good point.

jbteves replied:
Alternative of opening Draft PR and requesting help is proposed in 6aa2296

emdupre commented on CONTRIBUTING.md:
````diff
@@ -202,20 +237,45 @@ When opening a pull request, please use at least one of the following prefixes:
+```
+
+In the event of confusion, please ping in the Gitter for help.
+We strongly recommend you *not* use the `git rebase` command when working with and contributing to the `tedana` repository.
````
Can we drop the mention of rebase here, too ?

jbteves replied:
See other comment.

emdupre commented on CONTRIBUTING.md:
```diff
@@ -202,20 +237,45 @@ When opening a pull request, please use at least one of the following prefixes:
+In the event of confusion, please ping in the Gitter for help.
+We strongly recommend you *not* use the `git rebase` command when working with and contributing to the `tedana` repository.
+
+### Advanced Development
```
Unclear what this means to a naive reader. Can you include examples (e.g., `such as....`) ?

jbteves replied:
I'm not sure that I have a great definition here; perhaps we could say something like, "Comprehensive Developer Guidelines" instead?

jbteves replied:
I just renamed it "Comprehensive Developer Guidelines" in b7374a1

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+Unit tests focus on testing individual functions, whereas integration tests focus on making sure
+that the whole workflow runs correctly.
+
+For unit tests, we try to keep tests on the same module grouped into one file.
```
```suggestion
For unit tests, we try to keep tests from the same module grouped into one file.
```

emdupre replied:
I'd add a subheading to this to specify that it's unit testing (as opposed to other cases outlined below)

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+You may also use ``pytest.raises`` to ensure that errors are thrown for invalid inputs to a
+function.
+
+For integration tests, make a ``tar.gz`` file which will unzip to be only the files you'd like to
```
I would couch this -- it's going to be fairly uncommon at this point. I might say something like:

> In rare cases, you may need to add a new integration test. This means adding a new kind of dataset that tedana will be explicitly tested to support. As always, please open an issue first if you think that you have such a dataset ! Then, after making sure you have appropriate permission to share it, follow these steps to add it to the tedana repository.

emdupre replied:
Also, this should be a new section !

jbteves replied:
Addressed in 6a6a71f and 13e3daa

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+Using the five-echo set as a template, you should then edit ``.circlec/config.yml`` to add your
+test, calling the same name you define in the ``Makefile``.
+
+If you need to take a look at a failed test on CircleCI rather than locally, you can use the
```
New section here, please.

jbteves replied:
Done in c720db1

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+If you don't want all of the artifacts, you can go to the test details and use the browser to
+manually select the files you would like.
+
+Worked Example
```
Can we add pictures, here ??

jbteves replied:
What specifically would you like pictures for?

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+
+.. code-block:: bash
+
+    git checkout feature/say_hello  # unless you're already there
```
It won't hurt even if you are, right ?

jbteves replied:
That's correct, I could amend the wording to make that more clear if you like.

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+Once that test is passing, we may need to adjust the integration test.
+Our program creates a file, ``hello_world.txt``, which the older version would not have produced.
+Therefore, we need to add the file to ``$TEDANADIR/tedana/tests/data/tedana_outputs.txt`` and its
+counterpart, R2-D2-- uh, we mean, ``tedana_outputs_verbose.txt``.
```
:laughing:

emdupre commented on docs/developing.rst:
```diff
@@ -0,0 +1,199 @@
+
+and we're good to go!
+
+Monthly Developer Calls
```
I feel like this is sort of buried, here. Can we put it at the top of the page ?

jbteves replied:
Sure! I wasn't sure what priority this has versus other sections.

jbteves replied:
Fixed in 7102ff0

emdupre commented:
Also, I have it as a specific comment but it's general enough that I want to include it outside of the line edit:

I think the way this guide is currently written risks discouraging people from opening pull requests "early and often" which is something we explicitly want. I want to make sure we encourage folks to open PRs as soon as they get the go-ahead on an issue. It's ok if they're not fully documented or tested -- having the idea to look over helps give guidance and make sure someone isn't going off on the wrong path.

_But_ I do think it's important to set expectations that things won't be merged without a proper review, and they can't have a proper review without those docs, tests, etc !

jbteves commented on CONTRIBUTING.md:
````diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
+git merge upstream/master
+```
+
+If you know what rebasing is, please only use it for changes that haven't been pushed.
````
The problem I've run into is that new git users will look up how to do something, and then see a StackExchange post saying to rebase or some forum post saying to do so, and then they do so without realizing the consequences. For example, the Linux kernel frequently recommends a rebase, but they're also targeting users much more git-familiar than a typical contributor in open source science projects. This has happened a few times, so I would like to do something to indicate to people that we don't recommend it unless they're sure of what they're doing.

emdupre replied:
Maybe we could say something more generally about not amending history in pull requests ? So it would cover git commit --amend, too ? We can link here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

And point out that rewriting history can make it more difficult for others to work collaboratively with you on a new feature or bug fix ?

jbteves replied:
That makes sense. I always forget that you can `--amend` because usually I do my cleanup as an interactive rebase. I probably should have dropped the link in there to begin with, it's the link I use in my tutorials for my lab. I'll go ahead and work it into that section, thanks!

jbteves replied:
Do you like how it's handled in 1bd9b19?

jbteves replied:
I've opted to just delete this part since we like how it's handled in the other section. I figure it's somewhat unlikely for someone to attempt a rebase at that point unless they're aware of that version of the workflow.

jbteves commented on CONTRIBUTING.md:
````diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
+git checkout -b MYBRANCH
+```
+
+As you're making changes, make sure your branch is kept up to date with 
````
These usually end up getting erased as we squash and merge, though, right?

jbteves replied:
@emdupre do you mind commenting really quickly on this?

emdupre replied:
I'm leaning against it only because I think it's not necessary unless there's a merge conflict !

jbteves replied:
Hm, I see your point, I'll reword it. For novices that could be confusing.

jbteves commented on CONTRIBUTING.md:
````diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
 
 ```
-docker run --tty --rm -v ${PWD}:/tedana tedana/tedana-dev:latest run_all_tests
+# All tests; final checks before pushing
````
I'm not sure I follow, sorry, could you elaborate?

rmarkello replied:
I think (in line with @emdupre's comment) it would be great to take each of the `pytest` commands below and simply add them beneath each of the descriptive sentences above. For example:

> If you know a file will test your change, you can run only that test with:
> 
> ```
> pytest -k my_function $TEDANADIR/tedana/tests/test_file.py
> ```

and so on.

jbteves commented:
I can see your point. I think the problem was that I wrote it in response to concerns of already-active developers, without considering that this would also be what *new* contributors read. Let me mull over some of the wording tomorrow and make some changes ahead of the holiday. I've gone ahead and committed all of your suggestions.

jbteves commented on CONTRIBUTING.md:
```diff
@@ -151,30 +136,72 @@ Once you've run this, your repository should be set for most changes (i.e., you
-(This does require that you have a local install of [Docker](https://www.docker.com/products/docker-desktop).)
-You can run all the checks with:
+and unit/integration tests by running `pytest` (more details below).
+If you know a file will test your change, you can run only that test.
```
Addressed in 267bf66

jbteves commented:
Also, is there any way to make my responses to *not* start a review? That was not my intention. I don't know why I'm even allowed to review myself.

rmarkello commented on docs/developing.rst:
```diff
@@ -0,0 +1,224 @@
+
+.. code-block:: bash
+
+    git checkout feature/say_hello  # unless you're already there
```
```suggestion
    git checkout feature/say_hello
```

jbteves replied:
Hm, I did this because @emdupre had pointed that it's unclear that you only have to do this if you've left the branch. Do you think this adds to the confusion instead of clarifying?

jbteves commented:
@rmarkello I think most of your concerns are addressed here except for the comment I left for the checkout in `developing.rst`. I went ahead and opted to delete the section about fetching and merging, because if a merge conflict occurs and someone doesn't know how to handle it I figure we can discuss it on their PR, does that seem like a good solution?

rmarkello commented:
Thanks, @jbteves! I know we've all gone back-and-forth on this quite a bit so I appreciate your accommodating all the comments and requests!

Re the point in `developing.rst`: I guess I had interpreted @emdupre's comment slightly differently. That is, since there's no harm to running `git checkout <branch>` even if you're already on the branch, we should just recommend users do it regardless of where they are (and not potentially confused them with the comment). This is a pretty minor point so I'm happy to leave it as-is, unless @emdupre has a preference.

jbteves commented:
Just wanted to make sure I shouldn't add any commentary in addition. Okay, I think all points have been resolved now. 

No problem accommodating requests, it's important that the dev docs be airtight.

jbteves commented:
@KirstieJane any chance you or someone in your lab could take a look at this?

jbteves commented:
Awesome! :raised_hands: 
I'll wait a few days to hear from Kirstie, and I'll also need @rmarkello's approval before merging as well. 
You're welcome!

jbteves commented:
Sorry to stale your reviews, but I realized I need to address #479 and that didn't make it in before the approving reviews.

Islast commented on CONTRIBUTING.md:
````diff
@@ -151,30 +136,66 @@ Once you've run this, your repository should be set for most changes (i.e., you
+Before making a new branch, make sure your master is up to date.
 
-Using a new branch allows you to follow the standard "fork/branch/commit/pull-request/merge" GitHub workflow when making changes. [This guide][link_gitworkflow] provides a useful overview for this workflow.
+```
````
maybe it's worth adding a sentence to the effect of "you can do this by running the following in your command line" in front of this code snippet

jbteves commented on CONTRIBUTING.md:
```diff
@@ -151,30 +136,66 @@ Once you've run this, your repository should be set for most changes (i.e., you
+We've found that working on a [new branch][link_branches] for each issue makes it easier to keep your changes targeted.
+Using a new branch allows you to follow the standard GitHub workflow when making changes.
+[This guide][link_gitworkflow] provides a useful overview for this workflow.
+Before making a new branch, make sure your master is up to date.
```
```suggestion
Before making a new branch, make sure your master is up to date with the following commands:
```

jbteves replied:
@Islast how does that sound?

Islast replied:
perfect :+1:

Islast commented on CONTRIBUTING.md:
```diff
@@ -11,7 +11,6 @@ Here are some [instructions][link_signupinstructions].
 Already know what you're looking for in this guide? Jump to the following sections:
 
```
This is a very useful section, I think there are a couple of other points in the document readers might want to quickly navigate to, so I would add:
* [installing in editable mode](#Run-the-developer-setup)
* [testing and linting](#Test-your-changes)
* [updating the documentation after changes to the code](#Changes-to-documentation)

I would also link to the developer docs here with a brief description of the contents, say: 
See also our Developing Guidelines on [readthedocs][link_developing_rtd] for additional information, particularly on our monthly developer calls and writing tests for tedana.

Islast commented on docs/developing.rst:
```diff
@@ -0,0 +1,224 @@
+``tedana``'s codebase, in particular working with tests.
+The worked example also offers some guidelines on approaching testing when
+adding new functions.
+
```
I would add a link to your contributing guide somewhere up here

jbteves commented:
@Islast thanks for your review! For some reason it's not letting me request a re-review but hopefully you see this note; I think I addressed your concerns.

Islast commented:
@jbteves I see them, I think the changes you've made are perfect.

jbteves commented:
D'oh, sorry @Islast looks like it didn't render the re-request on my end but it sent it to you anyway. Sorry 😬 
@all-contributors please add @Islast for reviewing pull requests.
@rmarkello Once I have your approval we'll merge this.

jbteves commented:
@all-contributors add @Islast for review
